### PR TITLE
core: ModuleOp.ops, Region.ops and Block.ops are BlockOps

### PR DESCRIPTION
**Pull request by superlopuh**
This makes sure that all users migrate their code, with no implicit assumption that indexing is O(1). If the user wants random access into the operations, they need to create their own lists, otherwise use for loops or `next(iter(ops))` or whatever.

As far as users of these methods are concerned, the linked list change is an implementation detail of BlockOps, and won't need to update their code.

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
Would it make sense to rebase this on main? That would remove changes that were already committed.

**Comment by superlopuh** [comment]
Not sure why it failed with the diff, merged main

**Review by georgebisbas — COMMENTED** [review]
Let's not hurry up to merge this, will review within the day

**Review by math-fehr — APPROVED** [review]
:+1:, just add  some documentation on `BlockOps`.

File: tests/test_printer.py [hunk 1]
```diff
@@ -146,7 +146,9 @@ def test_op_message():
     module = parser.parse_module()
 
     diagnostic = Diagnostic()
-    diagnostic.add_message(module.ops[0], "Test message")
+    first_op = module.ops.first
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply tests/test_printer.py:149]: Yes, `ops[0]` doesn't work any more, there is no random-access indexing operator

File: tests/dialects/test_gpu.py [hunk 2]
```diff
@@ -166,7 +166,7 @@ def test_gpu_module():
     gpu_module = ModuleOp.get(name, ops)
 
     assert isinstance(gpu_module, ModuleOp)
-    assert gpu_module.body.ops == ops
+    assert list(gpu_module.body.ops) == ops
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply tests/dialects/test_gpu.py:169]: Yep

[the remark this answers is not in the document]
Comment (superlopuh) [reply tests/dialects/test_gpu.py:169]: Well `__eq__` is not defined on `BlockOps` is the real issue

**Review by georgebisbas — APPROVED** [review]
ok, thanks

**Pull request merged**